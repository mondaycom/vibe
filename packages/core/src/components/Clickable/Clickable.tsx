import cx from "classnames";
import React, { AriaRole, forwardRef } from "react";
import { noop as NOOP } from "lodash-es";
import VibeComponentProps from "../../types/VibeComponentProps";
import VibeComponent from "../../types/VibeComponent";
import useClickableProps from "../../hooks/useClickableProps/useClickableProps";
import styles from "./Clickable.module.scss";

export interface ClickableProps extends VibeComponentProps {
  /**
   * The HTML element or custom component used as the root.
   */
  elementType?: keyof JSX.IntrinsicElements | string;
  /**
   * The content inside the clickable element.
   */
  children?: React.ReactNode;
  /**
   * The ARIA role of the element.
   */
  role?: AriaRole;
  /**
   * Callback fired when the element is clicked.
   */
  onClick?: (event: React.MouseEvent | React.KeyboardEvent) => void;
  /**
   * If true, allows text selection inside the element.
   */
  enableTextSelection?: boolean;
  /**
   * Callback fired when the mouse button is pressed down on the element.
   */
  onMouseDown?: (event: React.MouseEvent) => void;
  /**
   * Callback fired when the mouse enters the element.
   */
  onMouseEnter?: (event: React.MouseEvent) => void;
  /**
   * Callback fired when the mouse leaves the element.
   */
  onMouseLeave?: (event: React.MouseEvent) => void;
  /**
   * The label of the element for accessibility.
   */
  ariaLabel?: string;
  /**
   * If true, hides the element from assistive technologies.
   */
  ariaHidden?: boolean;
  // TODO: [breaking] remove string type
  /**
   * Indicates the presence of a popup associated with the element.
   */
  ariaHasPopup?: boolean | string;
  /**
   * If true, indicates that the associated popup is open.
   */
  ariaExpanded?: boolean;
  // TODO: [breaking] remove string type
  /**
   * The tab order of the element.
   */
  tabIndex?: string | number;
  /**
   * If true, the element is disabled.
   */
  disabled?: boolean;
  /**
   * Inline styles applied to the element.
   */
  style?: React.CSSProperties;
}

const Clickable: VibeComponent<ClickableProps, HTMLElement> = forwardRef(
  (
    {
      elementType = "div",
      className = "",
      id,
      children,
      role = "button",
      onClick = NOOP,
      enableTextSelection = false,
      onMouseDown = NOOP,
      onMouseEnter = NOOP,
      onMouseLeave = NOOP,
      ariaLabel,
      ariaHidden,
      ariaHasPopup,
      ariaExpanded,
      tabIndex = "0",
      disabled = false,
      style,
      "data-testid": dataTestId
    }: ClickableProps,
    ref: React.ForwardedRef<HTMLElement>
  ) => {
    const clickableProps = useClickableProps(
      {
        onClick,
        onMouseDown,
        onMouseEnter,
        onMouseLeave,
        disabled,
        id,
        "data-testid": dataTestId,
        role,
        tabIndex,
        ariaLabel,
        ariaHidden,
        ariaHasPopup,
        ariaExpanded
      },
      ref
    );
    const overrideClassName = cx(styles.clickable, className, {
      [styles.disabled]: disabled,
      [styles.disableTextSelection]: !enableTextSelection
    });

    return React.createElement(
      elementType,
      {
        ...clickableProps,
        className: overrideClassName,
        style: style
      },
      children
    );
  }
);

export default Clickable;
