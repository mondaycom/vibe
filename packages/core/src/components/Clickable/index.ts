export { default as Clickable, type ClickableProps } from "./Clickable";
