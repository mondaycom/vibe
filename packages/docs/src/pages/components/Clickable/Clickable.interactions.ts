import { expect } from "@storybook/jest";
import {
  getByLabelText,
  pressNavigationKey,
  interactionSuite,
  NavigationCommand,
  resetFocus
} from "@vibe/core/interactionsTests";

async function states_onClickTabFocusElementTest(canvas) {
  const CLICKABLE_LABEL = "clickable button";
  const CLICKABLE_DISABLED_LABEL = "disabled clickable button";
  const clickableElement = await getByLabelText(canvas, CLICKABLE_LABEL);
  const disabledClickableElement = await getByLabelText(canvas, CLICKABLE_DISABLED_LABEL);
  await pressNavigationKey(NavigationCommand.TAB);
  expect(document.activeElement).toEqual(clickableElement);
  await pressNavigationKey(NavigationCommand.TAB);
  expect(document.activeElement).not.toEqual(disabledClickableElement);
  expect(canvas).not.toContain(document.activeElement);
}

export const statesPlaySuite = interactionSuite({
  tests: [states_onClickTabFocusElementTest],
  afterEach: async () => {
    await resetFocus();
  }
});
