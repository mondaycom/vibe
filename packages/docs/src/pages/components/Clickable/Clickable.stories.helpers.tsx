import React from "react";
import { StorybookLink, Tip } from "vibe-storybook-components";

export const TipHookSolution = () => (
  <Tip title="Check out our hook solution for this use case">
    {`If you'd like to set clickable functionality on a specific element inside your React component instead of using a
    wrapper, please, take a look on our `}
    <StorybookLink size={StorybookLink.sizes.SMALL} page="Hooks/useClickableProps">
      useClickableProps
    </StorybookLink>{" "}
    hook.
  </Tip>
);
