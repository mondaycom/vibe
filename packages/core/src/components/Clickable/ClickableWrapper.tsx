import React, { forwardRef } from "react";
import Clickable from "./Clickable";
import VibeComponentProps from "../../types/VibeComponentProps";
import VibeComponent from "../../types/VibeComponent";
import { ClickableProps } from "../Clickable/Clickable";

export interface ClickableWrapperProps extends VibeComponentProps {
  /**
   * The content inside the wrapper.
   */
  children: React.ReactNode;
  /**
   * If true, wraps the content in a `Clickable` component.
   */
  isClickable: boolean;
  /**
   * Props passed to the `Clickable` component.
   */
  clickableProps: ClickableProps;
}

const ClickableWrapper: VibeComponent<ClickableWrapperProps, HTMLElement> = forwardRef(
  ({ children, isClickable = true, clickableProps = {} }: ClickableWrapperProps, ref) => {
    if (!isClickable) {
      return <>{children}</>;
    }

    return (
      <Clickable {...clickableProps} ref={ref}>
        {children}
      </Clickable>
    );
  }
);

export default ClickableWrapper;
