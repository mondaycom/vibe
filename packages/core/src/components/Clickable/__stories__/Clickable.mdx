import { Meta } from "@storybook/blocks";
import {
  BUTTON,
  HIDDEN_TEXT,
  ICON_BUTTON
} from "../../../storybook/components/related-components/component-description-map";
import * as ClickableStories from "./Clickable.stories";
import { TipHookSolution } from "./Clickable.stories.helpers";

<Meta of={ClickableStories} />

# Clickable

- [Overview](#overview)
- [Props](#props)
- [Usage](#usage)
- [Feedback](#feedback)

## Overview

An accessibility helper component, this component simulates a button on non button elements

<Canvas of={ClickableStories.Overview} />

### Import

```js
import { Clickable } from "@vibe/core";
```

## Props

<PropsTable />

## Variants

### States

Clickable component supports two different states: regular state and disabled state.
The state only affects the component functionality (a user cannot interact with a disabled clickable component) and not the component appearance.
You can use the component className and style props to change the component appearance.

<Canvas of={ClickableStories.States} />

<TipHookSolution />

## Usage

<UsageGuidelines
  guidelines={[
    "When you can't use button, but need to provide keyboard intractability",
    "This component behaves like a button, treat it like one"
  ]}
/>

## Related components

<RelatedComponents componentsNames={[BUTTON, HIDDEN_TEXT, ICON_BUTTON]} />
