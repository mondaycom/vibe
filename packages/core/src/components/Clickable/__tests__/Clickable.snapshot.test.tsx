import { it, expect } from "vitest";
import React from "react";
import renderer from "react-test-renderer";
import Clickable from "../Clickable";

it("renders correctly with empty props", () => {
  const tree = renderer.create(<Clickable />).toJSON();
  expect(tree).toMatchSnapshot();
});

it("renders correctly with props", () => {
  const tree = renderer
    .create(
      <Clickable
        onClick={() => console.log("test")}
        className="monday-style_tests-class"
        role="banner"
        enableTextSelection
      >
        Children text
      </Clickable>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
