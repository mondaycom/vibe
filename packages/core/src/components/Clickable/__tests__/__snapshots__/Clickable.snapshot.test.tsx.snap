// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders correctly with empty props 1`] = `
<div
  className="clickable disableTextSelection"
  data-testid="clickable"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="button"
  tabIndex={0}
/>
`;

exports[`renders correctly with props 1`] = `
<div
  className="clickable monday-style_tests-class"
  data-testid="clickable"
  onClick={[Function]}
  onKeyDown={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  role="banner"
  tabIndex={0}
>
  Children text
</div>
`;
